web/src/pill_typeahead.ts

```diff
@@ -64,54 +89,76 @@ export function set_up($input, pills, opts) {
             }
             return source;
         },
-        highlighter_html(item, query) {
-            if (include_streams(query)) {
+        highlighter_html(item: TypeaheadItem, query: string): string {
+            if (include_streams(query) && item.type === "stream") {
                 return typeahead_helper.render_stream(item);
             }
 
-            if (include_user_groups && user_groups.is_user_group(item)) {
+            if (
+                include_user_groups &&
+                item.type === "user_group" &&
+                user_groups.is_user_group(item)
+            ) {
                 return typeahead_helper.render_user_group(item);
             }
 
             // After reaching this point, it is sure
             // that given item is a person. So this
             // handles `include_users` cases along with
             // default cases.
-            return typeahead_helper.render_person(item);
+            assert(item.type === "user");
+            return typeahead_helper.render_person({
+                ...item,
+                is_broadcast: undefined,
+            });
         },
-        matcher(item, query) {
+        matcher(item: TypeaheadItem, query: string): boolean {
             query = query.toLowerCase();
             query = query.replaceAll("\u00A0", " ");
 
-            if (include_streams(query)) {
+            if (include_streams(query) && item.type === "stream") {
                 query = query.trim().slice(1);
                 return item.name.toLowerCase().includes(query);
             }
 
             let matches = false;
-            if (include_user_groups) {
+            if (include_user_groups && item.type === "user_group") {
                 matches = matches || group_matcher(query, item);
             }
 
-            if (include_users) {
+            if (include_users && item.type === "user") {
                 matches = matches || person_matcher(query, item);
             }
             return matches;
         },
-        sorter(matches, query) {
+        sorter(matches: TypeaheadItem[], query: string): TypeaheadItem[] {
             if (include_streams(query)) {
-                return typeahead_helper.sort_streams(matches, query.trim().slice(1));
+                const stream_matches: StreamPillData[] = [];
+                for (const match of matches) {
+                    assert(match.type === "stream");
+                    stream_matches.push(match);
```

Comment:
Is there a shorter way of making a copy of an array asserting every item is a certain type?

Reply:
Not obviously, but this is probably fine in the scheme of things.